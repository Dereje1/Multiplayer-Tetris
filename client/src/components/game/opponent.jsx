import React from 'react';
import PropTypes from 'prop-types';
import './opponentdescription.css';
// connect to redux
import { connect } from 'react-redux';

import {
  SIMULATE_GAMEPLAY,
} from '../../constants';

import { socket } from '../../Actions/socket';
import { clearCanvas, drawRuble, drawBoundary, drawCells } from './scripts/canvas';

// custom components
import OpponentDescription from './opponentInfo';

// reads from store
const mapStateToProps = state => state;

// writes to store
class Opponent extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      gameState: {},
      status: ['', 0], // [status, data]
      playerPool: [],
      selfSocketId: '',
      opponent: {},
      levelsRaised: 0,
    };
    this.canvasOpponent = React.createRef();

    socket.on('CURRENT_POOL', pool => this.processPool(pool));
    socket.on('INVITATION_RECEIVED', invitedBy => this.processInvite(invitedBy));
    socket.on('GAME_DISCONNECTED', sId => this.disconnectGame(sId));
    socket.on('START_GAME', opp => this.processGameStart(opp));
    socket.on(SIMULATE_GAMEPLAY, oppGame => this.processGame(oppGame));
    socket.on('GAME_END', win => this.processGameEnd(win));
  }
  componentDidMount() {
    this.countGameover = 0;
    socket.emit('PLAYER_JOINED', JSON.stringify(this.props.user));
    console.log('Opponent Mounted!!');
  }

  componentDidUpdate(prevProps, prevState) {
    if (Object.keys(prevState.gameState).length) {
      if (prevState.gameState.points.totalLinesCleared !==
         this.state.gameState.points.totalLinesCleared) {
        this.processFloorRaise();
      }
    }
    if (prevState.status[0] !== this.state.status[0]) {
      if ((this.state.status[0] === 'Playing') || (this.state.status[0] === 'GameOver')) {
        this.props.onDisableExit(true);
      }
    }
  }

  componentWillUnmount() {
    socket.emit('COMPONENT_UNMOUNTED', 'opponent');
    socket.emit('disconnect', '');
  }

  setGame = () => {
    if (Object.keys(this.state.gameState).length) {
      // full deep copy of game state needed as object mutation becomes a problem
      const copyOfState = JSON.parse(JSON.stringify(this.state.gameState));
      if (this.state.status[0] === 'GameOver') {
        clearCanvas(this.canvasOpponentContext, copyOfState);
        return;
      }
      const canvasOpponent = this.canvasOpponent.current;
      canvasOpponent.style.backgroundColor = 'black';
      this.canvasOpponentContext = canvasOpponent.getContext('2d');
      copyOfState.activeShape.unitBlockSize /= 2;
      clearCanvas(this.canvasOpponentContext, copyOfState);
      drawBoundary(this.canvasOpponentContext, copyOfState);
      drawCells(this.canvasOpponentContext, copyOfState.activeShape, true);
      drawRuble(this.canvasOpponentContext, copyOfState, true);
    }
  };

  setUp = () => {
    const opponentsAvailable = !!this.state.playerPool.length;
    if (!opponentsAvailable) {
      this.setState({
        status: ['noopponents', 0],
      });
    } else {
      this.setState({
        status: ['opponents', this.state.playerPool.length],
      });
    }
  }
  setDifficulty = (val) => {
    this.props.onSetDifficulty(val);
  }

  processFloorRaise = () => {
    /*
    Difficulty                                 Description
    -----------------------------------------------------------------------------------
      1               After player clears 4 rows , floor is raised by 1 row on opponent
      2               After player clears 3 rows , floor is raised by 1 row on opponent
      3               After player clears 2 rows , floor is raised by 1 row on opponent
      4               After player clears 1 row  , floor is raised by 1 row on opponent
    */
    const { totalLinesCleared } = this.state.gameState.points;
    const difficultyMap = [[4, 1], [3, 2], [2, 3], [1, 4]]; // [[level, ]]
    const amountNeededForRaise = difficultyMap.filter(d => d[0] === this.props.difficulty)[0][1];
    const targetRaised = Math.floor(totalLinesCleared * (1 / amountNeededForRaise));
    const toRaise = targetRaised - this.state.levelsRaised;
    if (toRaise > 0) this.props.onFloorRaise(Number(toRaise));
    this.setState({
      levelsRaised: this.state.levelsRaised + toRaise,
    });
  }

  /* process socket-in-coming below */
  processPool = (poolData) => {
    const checkSelfSocketId = this.state.selfSocketId === '' ? poolData.self : this.state.selfSocketId;
    const nonSelfPoolData = poolData.pool.filter(p => p.socketId !== checkSelfSocketId);
    const playerChoices = [];
    nonSelfPoolData.forEach((sock, idx) => {
      if (idx < 5) playerChoices.push(sock);
    });

    this.setState({
      playerPool: playerChoices,
      selfSocketId: this.state.selfSocketId === '' ? poolData.self : this.state.selfSocketId,
    }, () => this.setUp());
  }

  disconnectGame = () => {
    console.log('A signal has come in that your Opponent Disconnected the Game!!');
    // add penalty in database for this in the future
    this.props.onGameOver(null);
  }

  processInvite = (host) => {
    const fullPlayerInfo = this.state.playerPool.filter(p => p.socketId === host.hostSocketId)[0];
    this.setState({
      status: ['Invite', [fullPlayerInfo, host.difficulty]],
    });
  }

  processGameStart = (opp) => {
    this.setState({
      opponent: opp,
    });
    let startCounter = 15;
    const gameStartId = setInterval(() => {
      this.setState({
        status: ['PreGame', startCounter],
      });
      startCounter -= 1;
      if (startCounter <= 0) {
        this.setState({
          status: ['Playing', null],
        });
        this.props.onGameEmit({ self: this.state.selfSocketId, opponnent: opp });
        this.props.onReset();
        clearInterval(gameStartId);
      }
    }, 1000);
  }

  processGame = (msg) => {
    if (this.state.status[0] === 'GameOver') {
      clearCanvas(this.canvasOpponentContext, this.state.gameState);
      return;
    }
    const copyOfState = JSON.parse(JSON.stringify(this.state));
    copyOfState.gameState = JSON.parse(msg);
    this.setState(copyOfState, () => this.setGame());
  }

  processGameEnd = (isWinner) => {
    this.props.onClearCanvas();
    this.countGameover += 1;
    if (this.countGameover > 1) return;
    // clearCanvas(this.canvasOpponentContext, this.state.gameState);
    // isWinner , true if client won
    const databaseEntry = isWinner ?
      {
        difficulty: this.props.difficulty,
        multiPlayer: true,
        players: [
          {
            name: this.props.user.displayName,
            _id: this.props.user._id,
            score: this.props.game.points.totalLinesCleared * 50,
            winner: isWinner,
          },
          {
            name: this.state.opponent.displayName,
            _id: this.state.opponent._id,
            score: this.state.gameState.points.totalLinesCleared * 50,
            winner: !isWinner,
          },
        ],
      }
      : null;

    let startCounter = 10;
    this.setState({
      status: ['GameOver', isWinner],
    }, () => this.props.onPause(true));
    console.log(`game over called ${this.countGameover}`);
    const gameEndId = setInterval(() => {
      startCounter -= 1;
      if (startCounter <= 0) {
        if (databaseEntry) this.props.onGameOver(databaseEntry);
        else this.props.onGameOver(null);
        this.props.onDisableExit(false);
        clearInterval(gameEndId);
        this.props.onReset(false);
      }
    }, 1000);
  };

  /* process socket-out-going below */
  requestInvite = (p) => {
    socket.emit('INVITATION_SENT', { hostSocketId: p, difficulty: this.props.difficulty });
  }

  acceptInvite = () => {
    this.props.onSetDifficulty(this.state.status[1][1]);
    this.setState({
      status: ['Loading', null],
    });
    socket.emit('INVITATION_ACCEPTED', [this.state.selfSocketId, this.state.status[1][0].socketId]);
  }

  /* done sockets */

  render() {
    return (
      <div className="opponentContainer">
        <OpponentDescription
          opponentState={this.state}
          difficulty={this.props.difficulty}
          setDifficulty={this.setDifficulty}
          requestInvite={sId => this.requestInvite(sId)}
          acceptInvite={() => this.acceptInvite()}
        />
        <canvas
          ref={this.canvasOpponent}
          width={this.props.game.canvas.canvasMajor.width / 2}
          height={this.props.game.canvas.canvasMajor.height / 2}
        />
      </div>
    );
  }

}

Opponent.defaultProps = {
  game: {},
  user: {},
  onFloorRaise: null,
  onReset: null,
  onGameEmit: null,
  onGameOver: null,
  onSetDifficulty: null,
  onClearCanvas: null,
  onPause: null,
  onDisableExit: null,
  difficulty: 2,
};
Opponent.propTypes = {
  game: PropTypes.objectOf(PropTypes.any),
  user: PropTypes.objectOf(PropTypes.any),
  difficulty: PropTypes.number,
  onFloorRaise: PropTypes.func,
  onReset: PropTypes.func,
  onGameEmit: PropTypes.func,
  onGameOver: PropTypes.func,
  onSetDifficulty: PropTypes.func,
  onClearCanvas: PropTypes.func,
  onPause: PropTypes.func,
  onDisableExit: PropTypes.func,
};

export default connect(mapStateToProps)(Opponent);

