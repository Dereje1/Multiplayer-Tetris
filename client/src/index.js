/* default CRA imports */
import React from 'react';
import ReactDOM from 'react-dom';
import './index.css';
/* custom imports */
import { Provider } from 'react-redux';
import store from './redux/store';
import Router from './routes/index';


ReactDOM.render(
  <Provider store={store}>
    <Router />
  </Provider>,
  document.getElementById('root'),
);
