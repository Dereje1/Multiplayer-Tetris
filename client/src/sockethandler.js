import io from 'socket.io-client';
import store from './redux/store';
import { socket as socketConstants } from './constants/index';
import {
  getLoggedInUsers, getClientSocketId, getOpponents, removeOpponents,
  sendInvite, receiveInvite, declinedInvitation, acceptedInvitation,
} from './redux/actions/socket';

const socketConnection = io(socketConstants.connection);
const {
  serverEmit: {
    LOGGED_IN_USERS, SOCKET_ID, SERVER_RESET, OPPONENT_POOL,
    UNMOUNT_OPPONENT, INVITE_SENT, INVITE_RECIEVED, DECLINED_INVITATION,
    ACCEPTED_INVITATION,
  },
} = socketConstants;
const { clientEmit: { SEND_LOGGED_IN_USER } } = socketConstants;

const clientEmitter = (event, dataToEmit) => {
  socketConnection.emit(event, dataToEmit);
};

const serverReset = () => {
  // runs whenever there is a new connection.
  // If client has not refreshed then use existing user profile to update server
  const { socket: socketObject, user } = store.getState();
  if (!Object.keys(socketObject).length) return;
  const payloadToSend = user.profile.authenticated ? user.profile : null;
  clientEmitter(
    SEND_LOGGED_IN_USER,
    payloadToSend,
  );
};

socketConnection.on(
  SERVER_RESET,
  () => serverReset(),
);

socketConnection.on(
  LOGGED_IN_USERS,
  data => store.dispatch(getLoggedInUsers(data)),
);

socketConnection.on(
  SOCKET_ID,
  data => store.dispatch(getClientSocketId(data)),
);

socketConnection.on(
  OPPONENT_POOL,
  data => store.dispatch(getOpponents(data)),
);

socketConnection.on(
  UNMOUNT_OPPONENT,
  () => store.dispatch(removeOpponents()),
);

socketConnection.on(
  INVITE_SENT,
  reciever => store.dispatch(sendInvite(reciever)),
);

socketConnection.on(
  INVITE_RECIEVED,
  sender => store.dispatch(receiveInvite(sender)),
);

socketConnection.on(
  DECLINED_INVITATION,
  () => store.dispatch(declinedInvitation()),
);

socketConnection.on(
  ACCEPTED_INVITATION,
  opponentData => store.dispatch(acceptedInvitation(opponentData)),
);

export default clientEmitter;
