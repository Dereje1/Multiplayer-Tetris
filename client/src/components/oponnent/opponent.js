import React from 'react';
import PropTypes from 'prop-types';
// connect to redux for store read onlny
import { connect } from 'react-redux';
import { clientEmitter } from '../../sockethandler';
import { socket as socketConstants } from '../../constants/index';
import { drawShape, drawBoundary } from '../game/scripts/canvas';
// custom components
import OpponentDescription from './opponentInfo';
import './styles/opponentdescription.scss';
import soundFile from './styles/WPN.wav';

// read from store
const mapStateToProps = ({ game, socket: { temp } }) => ({ game, temp });
const {
  clientEmit: {
    LOOK_FOR_OPPONENTS,
    OPPONENT_UNMOUNTED,
    INVITATION_SENT,
    INVITATION_DECLINED,
    START_GAME,
    UPDATED_CLIENT_SCREEN,
  },
} = socketConstants;

class Opponent extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      levelsRaised: 0, // storage for surplus floor levels by opponent
      opponentLinesCleared: 0,
      canvasLoaded: false, // load canvas only once
      docHidden: false,
    };
    this.canvasOpponent = React.createRef();
  }

  componentDidMount() {
    console.log('Opponent Mounted!!');
    const { temp } = this.props;
    if (!temp) clientEmitter(LOOK_FOR_OPPONENTS, null);
  }

  componentDidUpdate(prevProps) {
    const { canvasLoaded, opponentLinesCleared, docHidden } = this.state;
    const { game: prevGame, temp: prevTemp } = prevProps;
    const {
      game, onReset, toggleMultiplayer, onCanvasFocus,
      temp, onSetDifficulty,
    } = this.props;
    /* load Opponent Canvas */
    if (!canvasLoaded && this.canvasOpponent.current) {
      this.loadOpponentCanvas();
      this.setState({ canvasLoaded: true });
    }
    /* handle game phases here */
    if (temp) {
      const tempKey = Object.keys(temp)[0];
      switch (tempKey) {
        case 'acceptedInvitation': {
          if (!prevTemp.acceptedInvitation) {
            onReset({ reStart: false });
            if (this.canvasOpponentContext && !this.canvasOpponentContext.canvas.hidden) {
              this.canvasOpponentContext.canvas.hidden = true;
            }
            break;
          }
          if (docHidden) this.audio.play();
          const { acceptedInvitation: { countdown: prevCountdown } } = prevTemp;
          const { acceptedInvitation: { countdown, difficulty } } = temp;
          const cdReachedZero = prevCountdown === 1 && countdown === 0;
          if (cdReachedZero) {
            this.audio.play();
            clientEmitter(START_GAME, {
              opponentInfo: temp[tempKey],
              clientScreen: JSON.stringify(game),
            });
            onSetDifficulty(difficulty);
            this.setState({ levelsRaised: 0, opponentLinesCleared: 0 });
          }
        }
          break;
        case 'gameInProgress': {
          const { gameInProgress: { info, opponentScreen } } = temp;
          if (!prevTemp.gameInProgress) { // game started
            onCanvasFocus();
            onReset({});
            toggleMultiplayer();
          } else { // game running
            /* Important
            It's is not enough to emit every time
            component updates, must emit when only the game
            actually CHANGES!!, otherwise big performance
            degredation if we have other things, like setstate
            update the component, meaning a new emit on every component
            update!
            */
            const { gameInProgress: { opponentScreen: prevOpponentScreen } } = prevTemp;
            // set opponent screen on socket data only if there is a difference in the opp game.
            if (opponentScreen !== prevOpponentScreen) {
              this.setGame(opponentScreen, prevOpponentScreen);
            }
            // emit client data only if there is a difference in client game.
            if (JSON.stringify(prevGame) === JSON.stringify(game)) return;
            clientEmitter(UPDATED_CLIENT_SCREEN, {
              opponentSID: info.opponentSID,
              clientScreen: JSON.stringify(game),
            });
          }
        }
          break;
        case 'gameOver': {
          const { onGameOver } = this.props;
          if (!prevTemp.gameOver) {
            onGameOver([opponentLinesCleared]);
            toggleMultiplayer();
          }
        }
          break;
        default:
          break;
      }
    }
  }

  componentWillUnmount() {
    const { temp } = this.props;
    // if a person unmounts in the middle of an invitation
    if (temp && temp.invitationFrom) clientEmitter(INVITATION_DECLINED, temp);
    clientEmitter(OPPONENT_UNMOUNTED, temp);
  }

  loadOpponentCanvas = () => {
    // load only once
    const canvasOpponent = this.canvasOpponent.current;
    canvasOpponent.style.backgroundColor = 'black';
    this.canvasOpponentContext = canvasOpponent.getContext('2d');
    this.canvasOpponentContext.canvas.hidden = true;
  };

  // called on every game object change
  setGame = (opponentScreen, prevOpponentScreen) => {
    if (!opponentScreen) return;
    const { temp } = this.props;
    const opp = JSON.parse(opponentScreen);
    const prevOpp = prevOpponentScreen ? JSON.parse(prevOpponentScreen) : null;
    if (temp.gameOver) return;
    // test if a floor raise is warranted
    if (opp && prevOpp) this.processFloorRaise(opp, prevOpp);
    if (this.canvasOpponentContext.canvas.hidden) this.canvasOpponentContext.canvas.hidden = false;
    opp.activeShape.unitBlockSize /= 2;
    drawShape(this.canvasOpponentContext, opp, true);
  };

  setDifficulty = (val) => {
    const { onSetDifficulty } = this.props;
    onSetDifficulty(val);
  };

  // test if a floor raise is warranted
  processFloorRaise = (currentGame, previousGame) => {
    const { levelsRaised } = this.state;
    const {
      temp:
          {
            gameInProgress:
            {
              info: { difficulty },
            },
          },
      onFloorRaise, floorsRaisedOnOpp,
    } = this.props;
    const {
      points: { totalLinesCleared: previouslyClearedLines },
      rubble: { boundaryCells: prevBoundryCells },
    } = previousGame;
    const {
      points: { totalLinesCleared },
      rubble: { boundaryCells },
    } = currentGame;
    // draw boundry in opponent screen if floor raise
    if (boundaryCells.length !== prevBoundryCells.length) {
      const copyOfGame = JSON.parse(JSON.stringify(currentGame));
      copyOfGame.activeShape.unitBlockSize /= 2;
      drawBoundary(this.canvasOpponentContext, copyOfGame, true);
      floorsRaisedOnOpp((boundaryCells.length - prevBoundryCells.length) / 10);
    }
    const linesCleared = totalLinesCleared - previouslyClearedLines;
    // return if no new lines have been cleared
    if (!linesCleared) return;
    this.setState({ opponentLinesCleared: totalLinesCleared });
    /*
    Difficulty                                 Description
    -----------------------------------------------------------------------------------
      1               After player clears 4 rows , floor is raised by 1 row on opponent
      2               After player clears 3 rows , floor is raised by 1 row on opponent
      3               After player clears 2 rows , floor is raised by 1 row on opponent
      4               After player clears 1 row  , floor is raised by 1 row on opponent
    */
    const difficultyMap = [[4, 1], [3, 2], [2, 3], [1, 4]]; // [[level, ]]
    // number of floors that needs to be cleared for a single floor raise on opp
    const amountNeededForRaise = difficultyMap.filter(d => d[0] === difficulty)[0][1];
    // Includes any surplus from previous lines cleared
    const totalRaisedByOpponent = levelsRaised + linesCleared;
    if (totalRaisedByOpponent >= amountNeededForRaise) {
      // Total levels to be raised on opponent
      const raiseOnClient = Math.floor(totalRaisedByOpponent / amountNeededForRaise);
      // To store for client if any surplus
      const storeForOpponent = totalRaisedByOpponent - (raiseOnClient * amountNeededForRaise);
      this.setState({ levelsRaised: storeForOpponent }, () => onFloorRaise(Number(raiseOnClient)));
    } else this.setState({ levelsRaised: totalRaisedByOpponent });
  };

  /* process socket-out-going below */
  requestInvite = (sentTo) => {
    const { difficulty } = this.props;
    clientEmitter(INVITATION_SENT, { sentTo, difficulty });
  };

  resetMultiplayer = () => {
    const { onReset } = this.props;
    onReset({ reStart: false });
    clientEmitter(LOOK_FOR_OPPONENTS, null);
    if (!this.canvasOpponentContext.canvas.hidden) this.canvasOpponentContext.canvas.hidden = true;
  };
  /* done sockets */

  audioPlayer = () => (
    <audio ref={(input) => { this.audio = input; }} src={soundFile}>
      <track kind="captions" />
    </audio>
  );

  render() {
    const { difficulty, game, temp } = this.props;
    if (!temp) return null;
    return (
      <div className="opponentContainer">
        <OpponentDescription
          socketState={temp}
          difficulty={difficulty}
          setDifficulty={this.setDifficulty}
          requestInvite={sId => this.requestInvite(sId)}
          acceptInvite={() => this.acceptInvite()}
          declineInvite={() => this.declineInvite()}
          getPool={() => this.resetMultiplayer()}
        />
        <canvas
          ref={this.canvasOpponent}
          width={game.canvas.canvasMajor.width / 2}
          height={game.canvas.canvasMajor.height / 2}
        />
        {this.audioPlayer()}
      </div>
    );
  }

}

Opponent.defaultProps = {
  game: {}, // client game in redux store
  temp: null, // socket info in redux store
  onFloorRaise: null,
  onReset: null, // callback to main game
  onSetDifficulty: null,
  difficulty: 2,
  onGameOver: null,
  toggleMultiplayer: null,
  onCanvasFocus: null,
  floorsRaisedOnOpp: null,
};
Opponent.propTypes = {
  temp: PropTypes.objectOf(PropTypes.any),
  game: PropTypes.objectOf(PropTypes.any),
  difficulty: PropTypes.number,
  onFloorRaise: PropTypes.func,
  onReset: PropTypes.func,
  onSetDifficulty: PropTypes.func,
  onGameOver: PropTypes.func,
  toggleMultiplayer: PropTypes.func,
  onCanvasFocus: PropTypes.func,
  floorsRaisedOnOpp: PropTypes.func,
};

export default connect(mapStateToProps)(Opponent);
