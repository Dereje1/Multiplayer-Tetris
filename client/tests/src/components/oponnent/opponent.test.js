import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Opponent } from '../../../../src/components/oponnent/opponent';
import { gameStub } from '../../../stub'
import * as socketActions from '../../../../src/sockethandler'
import * as canvas from '../../../../src/components/game/scripts/canvas';

const getRefSpy = () => jest.spyOn(React, "createRef").mockImplementation(() => ({
    current: {
        style: {},
        getContext: () => ({
            canvas: {
                hidden: false
            },
            clearRect: jest.fn(),
            fillRect: jest.fn(),
            beginPath: jest.fn(),
            rect: jest.fn(),
            stroke: jest.fn(),
        })
    }
}))
describe('The Opponent component', () => {
    let props
    let useRefSpy;
    let socketSpy;
    beforeEach(() => {
        props = {
            game: gameStub,
            difficulty: 2,
            temp: {},
            onSetDifficulty: jest.fn(),
            onReset: jest.fn()
        }
        useRefSpy = getRefSpy();
        socketSpy = jest.spyOn(socketActions, 'clientEmitter');
    })
    afterEach(() => {
        props = null
        useRefSpy = null
        socketSpy.mockClear()
    })
    test('will render the oponent window', () => {
        const wrapper = shallow(<Opponent {...props} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    })

    test('will look for oponents on CDM', () => {
        delete props.temp
        shallow(<Opponent {...props} />);
        expect(socketSpy).toHaveBeenCalledWith("LOOK_FOR_OPPONENTS", null)
    })

    test('will set the difficulty level', () => {
        const wrapper = shallow(<Opponent {...props} />);
        const OpponentDescription = wrapper.find('OpponentDescription')
        OpponentDescription.props().setDifficulty(3)
        expect(props.onSetDifficulty).toHaveBeenCalledWith(3)
    })

    test('will request an invite', () => {
        const wrapper = shallow(<Opponent {...props} />);

        const OpponentDescription = wrapper.find('OpponentDescription')
        OpponentDescription.props().requestInvite('stubSocketId')
        expect(socketSpy).toHaveBeenCalledWith("INVITATION_SENT", { "difficulty": 2, "sentTo": "stubSocketId" })
    })

    test('will reset the connection', () => {
        const wrapper = shallow(<Opponent {...props} />);
        wrapper.instance().loadOpponentCanvas()
        const OpponentDescription = wrapper.find('OpponentDescription')
        OpponentDescription.props().getPool()
        expect(props.onReset).toHaveBeenCalledWith({ reStart: false })
        expect(socketSpy).toHaveBeenCalledWith("LOOK_FOR_OPPONENTS", null)
    })
})

describe('Component updates (CDU, CWU)', () => {
    let props
    let useRefSpy;
    let socketSpy;
    beforeEach(() => {
        props = {
            game: gameStub,
            difficulty: 2,
            temp: {},
            onSetDifficulty: jest.fn(),
            onReset: jest.fn(),
            onCanvasFocus: jest.fn(),
            toggleMultiplayer: jest.fn(),
            onGameOver: jest.fn()
        }
        useRefSpy = getRefSpy();
        socketSpy = jest.spyOn(socketActions, 'clientEmitter');
    })
    afterEach(() => {
        props = null
        useRefSpy = null
        socketSpy.mockClear()
    })

    test('will reset on a new invite accepted', () => {
        const prevProps = { game: {}, temp: {} };
        const updatedProps = {
            ...props,
            temp: {
                acceptedInvitation: {}
            }
        }
        const wrapper = shallow(<Opponent {...updatedProps} />);
        wrapper.instance().componentDidUpdate(prevProps)
        expect(props.onReset).toHaveBeenCalledWith({ reStart: false })
    })

    test('will start game when countdown reaches 0', () => {
        const prevProps = { game: {}, temp: { acceptedInvitation: { countdown: 1 } } };
        const updatedProps = {
            ...props,
            temp: {
                acceptedInvitation: { countdown: 0, difficulty: 4 }
            }
        }
        const wrapper = shallow(<Opponent {...updatedProps} />);
        wrapper.instance().audio = { play: jest.fn() }
        wrapper.instance().componentDidUpdate(prevProps)
        expect(socketSpy).toHaveBeenCalledWith("START_GAME", {
            clientScreen: JSON.stringify(gameStub),
            opponentInfo: { countdown: 0, difficulty: 4 }
        })
        expect(props.onSetDifficulty).toHaveBeenCalledWith(4)
    })

    test('will handle game in progress when game has not started yet', () => {
        const prevProps = { game: {}, temp: {} };
        const updatedProps = {
            ...props,
            temp: {
                gameInProgress: {}
            }
        }
        const wrapper = shallow(<Opponent {...updatedProps} />);
        wrapper.instance().componentDidUpdate(prevProps)
        expect(props.onCanvasFocus).toHaveBeenCalled()
        expect(props.onReset).toHaveBeenCalledWith({})
        expect(props.toggleMultiplayer).toHaveBeenCalled()
    })

    test('will do nothing if game is in progress but both client and opponent values have not changed', () => {
        const prevProps = {
            game: { ...gameStub },
            temp: {
                gameInProgress: { opponentScreen: JSON.stringify({}) }
            }
        };
        const updatedProps = {
            ...props,
            temp: {
                gameInProgress: { opponentScreen: JSON.stringify({}) }
            },
        }
        const wrapper = shallow(<Opponent {...updatedProps} />);
        const ans = wrapper.instance().componentDidUpdate(prevProps)
        expect(ans).toBe(undefined)
    })

    test('will update game on opponent screen if values have changed', () => {
        const prevProps = {
            game: { ...gameStub },
            temp: {
                gameInProgress: { opponentScreen: JSON.stringify({ a: 1 }) }
            }
        };
        const updatedProps = {
            ...props,
            temp: {
                gameInProgress: { opponentScreen: JSON.stringify({}) }
            },
        }
        const wrapper = shallow(<Opponent {...updatedProps} />);
        wrapper.instance().setGame = jest.fn()
        const ans = wrapper.instance().componentDidUpdate(prevProps)
        expect(wrapper.instance().setGame).toHaveBeenCalledWith("{}", "{\"a\":1}")
        expect(ans).toBe(undefined)
    })

    test('will emit updated client screen if game has changed', () => {
        const prevProps = {
            game: { ...gameStub },
            temp: {
                gameInProgress: { opponentScreen: JSON.stringify({}) }
            }
        };
        const updatedProps = {
            ...props,
            temp: {
                gameInProgress: { opponentScreen: JSON.stringify({}), info: { opponentSID: 'stubSID' } }
            },
            game: {
                ...gameStub,
                activeShape: {
                    ...gameStub.activeShape,
                    yPosition: 30
                }
            }
        }
        const wrapper = shallow(<Opponent {...updatedProps} />);
        wrapper.instance().componentDidUpdate(prevProps)
        expect(socketSpy).toHaveBeenCalledWith("UPDATED_CLIENT_SCREEN", {
            clientScreen: JSON.stringify(updatedProps.game),
            opponentSID: 'stubSID'
        })
    })

    test('will handle game over', () => {
        const prevProps = { game: {}, temp: {} };
        const updatedProps = {
            ...props,
            temp: {
                gameOver: {}
            }
        }
        const wrapper = shallow(<Opponent {...updatedProps} />);
        wrapper.setState({ opponentLinesCleared: 5 })
        wrapper.instance().componentDidUpdate(prevProps)
        expect(props.onGameOver).toHaveBeenCalledWith([5])
        expect(props.toggleMultiplayer).toHaveBeenCalled()
    })

    test('will emit on component unmounting', () => {
        const wrapper = shallow(<Opponent {...props} />);
        wrapper.instance().componentWillUnmount()
        expect(socketSpy).toHaveBeenCalledWith("OPPONENT_UNMOUNTED", {})
    })

    test('will emit on component unmounting in middle of an invitation', () => {
        const updatedProps = {
            ...props,
            temp: {
                invitationFrom: {}
            }
        }
        const wrapper = shallow(<Opponent {...updatedProps} />);
        wrapper.instance().componentWillUnmount()
        expect(socketSpy).toHaveBeenNthCalledWith(1, "INVITATION_DECLINED", { invitationFrom: {} })
        expect(socketSpy).toHaveBeenNthCalledWith(2, "OPPONENT_UNMOUNTED", { invitationFrom: {} })
    })

})

describe('Updating the opponent screen/canvas', () => {
    let props
    let useRefSpy, socketSpy, refreshCanvasSpy;
    beforeEach(() => {
        props = {
            game: gameStub,
            difficulty: 2,
            temp: {},
            onSetDifficulty: jest.fn(),
            onReset: jest.fn()
        }
        useRefSpy = getRefSpy();
        socketSpy = jest.spyOn(socketActions, 'clientEmitter');
        refreshCanvasSpy = jest.spyOn(canvas, 'refreshCanvas');
    })
    afterEach(() => {
        props = null
        useRefSpy = null
        socketSpy.mockClear()
        refreshCanvasSpy.mockClear()
    })
    test('will do nothing when no opponent screen info', () => {
        const wrapper = shallow(<Opponent {...props} />);
        const ans = wrapper.instance().setGame(null, {})
        expect(ans).toBe(undefined)
    })

    test('will do nothing on game over', () => {
        const updatedProps = {
            ...props,
            temp: {
                gameOver: {}
            }
        }
        const wrapper = shallow(<Opponent {...updatedProps} />);
        const ans = wrapper.instance().setGame(JSON.stringify({}), null)
        expect(ans).toBe(undefined)
    })
    test('will process line cleared by opponent', () => {
        const updatedProps = {
            ...props,
            temp: {
                gameInProgress:
                {
                    info: { difficulty: 2 },
                },
            }
        }
        
        const wrapper = shallow(<Opponent {...updatedProps} />);
        wrapper.instance().loadOpponentCanvas();
        expect(wrapper.state().opponentLinesCleared).toBe(0);
        expect(wrapper.state().levelsRaised).toBe(0);
        wrapper.instance().setGame(JSON.stringify({
            ...gameStub,
            points: {
                ...gameStub.points,
                totalLinesCleared: 1
            }
        }), JSON.stringify({ ...gameStub }))
        expect(wrapper.state().opponentLinesCleared).toBe(1);
        expect(wrapper.state().levelsRaised).toBe(1);
        expect(refreshCanvasSpy).toHaveBeenCalledTimes(1)
    })

    test('will draw floor raise on opponent if diff in boundary', () => {
        const updatedProps = {
            ...props,
            temp: {
                gameInProgress:
                {
                    info: { difficulty: 2 },
                },
            },
            floorsRaisedOnOpp: jest.fn()
        }
        
        const wrapper = shallow(<Opponent {...updatedProps} />);
        wrapper.instance().loadOpponentCanvas();
        expect(wrapper.state().opponentLinesCleared).toBe(0);
        expect(wrapper.state().levelsRaised).toBe(0);
        wrapper.instance().setGame(JSON.stringify({
            ...gameStub,
            points: {
                ...gameStub.points,
                totalLinesCleared: 1
            },
            rubble:{
                ...gameStub.rubble,
            },
            floor:{
                floorHeight: 1,
                floorIndices: []
            }
        }), JSON.stringify({ ...gameStub }))
        expect(wrapper.state().opponentLinesCleared).toBe(1);
        expect(wrapper.state().levelsRaised).toBe(1);
        expect(refreshCanvasSpy).toHaveBeenCalledTimes(1)
        expect(updatedProps.floorsRaisedOnOpp).toHaveBeenCalledWith(1)
    })
})