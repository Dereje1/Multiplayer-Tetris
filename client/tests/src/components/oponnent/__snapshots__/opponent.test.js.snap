// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Opponent component will render the oponent window 1`] = `
<div
  className="opponentContainer"
>
  <OpponentDescription
    difficulty={2}
    getPool={[Function]}
    requestInvite={[Function]}
    setDifficulty={[Function]}
    socketState={Object {}}
  />
  <canvas
    height={300}
    width={150}
  />
  <audio
    src={
      Object {
        "canInstrument": true,
        "createTransformer": [Function],
        "getCacheKey": [Function],
        "getCacheKeyAsync": [Function],
        "process": [Function],
        "processAsync": [Function],
      }
    }
  >
    <track
      kind="captions"
    />
  </audio>
</div>
`;
